From stable-bounces@linux.kernel.org  Tue May 31 07:08:57 2005
Date: Tue, 31 May 2005 16:08:47 +0200
From: Andi Kleen <ak@suse.de>
To: stable@kernel.org
ubject: [PATCH] x86_64: avoid SMP boot up race

Keep interrupts disabled during smp bootup

This avoids a race that breaks SMP bootup on some machines.
The race is not fully plugged (that is only done with much
more changes in 2.6.12), but should be good enough
for most people.

Keeping the interrupts disabled here is ok because we
don't rely on the timer interrupt for local APIC
timer setup, but always read the timer registers
directly.

(originally from Rusty Russell iirc) 

Signed-off-by: ak@suse.de
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/x86_64/kernel/apic.c    |    2 --
 arch/x86_64/kernel/smpboot.c |    4 ----
 2 files changed, 6 deletions(-)

--- linux-2.6.11.10.orig/arch/x86_64/kernel/apic.c	2005-05-16 10:50:31.000000000 -0700
+++ linux-2.6.11.10/arch/x86_64/kernel/apic.c	2005-06-08 22:40:23.000000000 -0700
@@ -775,9 +775,7 @@
 
 void __init setup_secondary_APIC_clock(void)
 {
-	local_irq_disable(); /* FIXME: Do we need this? --RR */
 	setup_APIC_timer(calibration_result);
-	local_irq_enable();
 }
 
 void __init disable_APIC_timer(void)
--- linux-2.6.11.10.orig/arch/x86_64/kernel/smpboot.c	2005-05-16 10:50:31.000000000 -0700
+++ linux-2.6.11.10/arch/x86_64/kernel/smpboot.c	2005-06-08 22:40:23.000000000 -0700
@@ -309,8 +309,6 @@
 	Dprintk("CALLIN, before setup_local_APIC().\n");
 	setup_local_APIC();
 
-	local_irq_enable();
-
 	/*
 	 * Get our bogomips.
 	 */
@@ -324,8 +322,6 @@
 	 */
  	smp_store_cpu_info(cpuid);
 
-	local_irq_disable();
-
 	/*
 	 * Allow the master to continue.
 	 */
